client/src/controllers/SyncController.ts

```diff
@@ -90,6 +90,59 @@ export class SyncController extends Controller<HTMLInputElement> {
     });
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  prefillTitleFromFilename(_event?: Event) {
+    const input = this.element;
+
+    // Find the target input via the selector in data-w-sync-target-value
+    const titleInput = document.querySelector<HTMLInputElement>(
+      this.targetValue,
+    );
+
+    if (!titleInput) {
+      return;
+    }
+
+    // Do not override if user has already typed a title
+    if (titleInput.value) return;
+
+    const rawPath = input.value || '';
+    const parts = rawPath.split('\\');
+    const filenameWithExt = parts[parts.length - 1] || '';
+
+    // Remove file extension
+    const title = filenameWithExt.replace(/\.[^.]+$/, '');
+
+    // Prepare details for the custom event
+    const maxTitleLengthAttr = titleInput.getAttribute('maxLength');
+    const maxTitleLength =
+      maxTitleLengthAttr !== null ? parseInt(maxTitleLengthAttr, 10) : null;
+
+    const customEvent = new CustomEvent('wagtail:images-upload', {
+      bubbles: true,
+      cancelable: true,
+      detail: {
+        data: { title },
+        filename: filenameWithExt,
+        maxTitleLength,
+      },
+    });
+
+    // Dispatch the event on the closest <form> element
+    const form = input.closest('form');
+    const cancelled = form ? !form.dispatchEvent(customEvent) : false;
+
+    if (cancelled) return;
```

Comment:
I think we should check for `defaultPrevented` specifically instead of the `dispatchEvent`'s return value so that it's easier to refactor this into a Stimulus `this.dispatch()` call (which returns the event object rather than whether it was cancelled).

```suggestion
    // Dispatch the event on the <form> element
    input.form.dispatchEvent(customEvent);
    if (customEvent.defaultPrevented) return;
```